src/lightning/fabric/utilities/optimizer.py

```diff
@@ -12,15 +12,36 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-from typing import Iterable
+import contextlib
+import functools
+import threading
+from typing import Callable, Dict, Generator, Iterable, TYPE_CHECKING
 
+import torch
 from lightning_utilities.core.apply_func import apply_to_collection
-from torch import Tensor
 from torch.optim import Optimizer
 
 from lightning.fabric.utilities.apply_func import move_data_to_device
+from lightning.fabric.utilities.imports import _TORCH_GREATER_EQUAL_2_0
 from lightning.fabric.utilities.types import _DEVICE
 
+if TYPE_CHECKING:
+    from lightning.fabric.wrappers import _FabricModule
+
+SUPPORTS_OPTIMIZER_IN_FSDP_BACKWARD = False
+if _TORCH_GREATER_EQUAL_2_0:
+    try:
+        from torch.distributed.fsdp._common_utils import _get_module_fsdp_state
+        from torch.distributed.fsdp._traversal_utils import _get_fsdp_handles
+        from torch.distributed.fsdp.flat_param import FlatParameter, FlatParamHandle
+
+        SUPPORTS_OPTIMIZER_IN_FSDP_BACKWARD = True
+
+    # Note: Not `ImportError` because sometimes PyTorch has internal failures
+    # (such as AttributeError) if it was built without distributed support.
+    except BaseException:
+        pass
```

Comment:
```suggestion

    # Note: Not `ImportError` because sometimes PyTorch has internal failures
    # (such as AttributeError) if it was built without distributed support.
    with contextlib.suppress(BaseException):
        from torch.distributed.fsdp._common_utils import _get_module_fsdp_state
        from torch.distributed.fsdp._traversal_utils import _get_fsdp_handles
        from torch.distributed.fsdp.flat_param import FlatParameter, FlatParamHandle

        SUPPORTS_OPTIMIZER_IN_FSDP_BACKWARD = True
```